### ISSUE TITLE
row_cache may get inconsistent when out of order writes and deletes are involved

### ISSUE THREAD

**Issue by fee-mendes** [scylladb/scylladb#20004]
**Installation details**
Scylla version (or git commit hash): `6.0.2-0.20240703.c9cd171f426e` with build-id `a8d5e83478cdc6b426ca045bb2d91992eadc7fda`

During cache merge (after a flush), rows with lower timestamp than already existing cache rows will be ignored, and may introduce an inconsistency between the row_cache and on-disk representation. 

This is more likely to occur:

- On tables with small write activity;
- On popular keys, given our LRU eviction policy
- When out of order writes are mixed with deletes
- When `tombstone_gc = {'mode': 'repair'}` is used, see `propagation_delay_in_seconds` or with a small `gc_grace_seconds` setting

This is a timing issue, and the results will vary depending in the order each step is carried out by each replica. Here's one example.

Consider `n1` and `n2`, and the following schema:

```cql
CREATE KEYSPACE k WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2};
CREATE TABLE k.t (
    pk int PRIMARY KEY,
    val int
) WITH tombstone_gc = {'mode': 'repair', 'propagation_delay_in_seconds': 20};
```

Consider this set of operations:

1. Delete a key.

```cql
CONSISTENCY ALL;
DELETE FROM k.t WHERE pk=0;
```

2. Flush `n1` and `n2`
3. Wait for `propagation_delay_in_seconds` to elapse and Repair
4. Run a major compaction on `n1`

At this point, `n1` will have evicted the tombstone. `n2` will still hold it on disk.

So far so good, all nodes agree `pk=0` is dead. 

1. Insert `pk=0` again using an out of order write. In this example, the key timestamp is lower than on the delete:

```cql
INSERT INTO k.t (pk, val) VALUES (0, 1) USING TIMESTAMP 1;
```

2. Flush `n1` and `n2`.

As expected, no surprises here. `n1` will mark `pk=0` as alive on disk. `n2` will contain nothing, since the 2nd flush evicted the tombstone, which was already expired given the previous repair.

The problem starts now:

1. Compare the `MUTATION_FRAGMENTS` from both `n1` and `n2`:

**N1**

```cql
 pk | mutation_source                                                                                               | partition_region | position_weight | metadata                                                                                     | mutation_fragment_kind | value
----+---------------------------------------------------------------------------------------------------------------+------------------+-----------------+----------------------------------------------------------------------------------------------+------------------------+-------------
  0 |                                                                                                     row-cache |                0 |                 |          {"tombstone":{"timestamp":1722725124073792,"deletion_time":"2024-08-03 22:45:24z"}} |        partition start |        null
  0 |                                                                                                     row-cache |                2 |                 | {"marker":{"timestamp":1},"columns":{"val":{"is_live":true,"type":"regular","timestamp":1}}} |         clustering row | {"val":"1"}
  0 |                                                                                                     row-cache |                3 |                 |                                                                                         null |          partition end |        null
  0 | sstable:/var/lib/scylla/data/k/t-b78e4b8051d311efa5814566d335f6f3/me-3gid_1rbj_24qcg2ilv31agwroxf-big-Data.db |                0 |                 |                                                                             {"tombstone":{}} |        partition start |        null
  0 | sstable:/var/lib/scylla/data/k/t-b78e4b8051d311efa5814566d335f6f3/me-3gid_1rbj_24qcg2ilv31agwroxf-big-Data.db |                2 |                 | {"marker":{"timestamp":1},"columns":{"val":{"is_live":true,"type":"regular","timestamp":1}}} |         clustering row | {"val":"1"}
  0 | sstable:/var/lib/scylla/data/k/t-b78e4b8051d311efa5814566d335f6f3/me-3gid_1rbj_24qcg2ilv31agwroxf-big-Data.db |                3 |                 |                                                                                         null |          partition end |        null

(6 rows)
```

**N2**
```cql
cqlsh> SELECT * FROM MUTATION_FRAGMENTS(k.t) WHERE pk=0;

 pk | mutation_source | partition_region | position_weight | metadata                                                                                     | mutation_fragment_kind | value
----+-----------------+------------------+-----------------+----------------------------------------------------------------------------------------------+------------------------+-------------
  0 |       row-cache |                0 |                 |          {"tombstone":{"timestamp":1722725124073792,"deletion_time":"2024-08-03 22:45:24z"}} |        partition start |        null
  0 |       row-cache |                2 |                 | {"marker":{"timestamp":1},"columns":{"val":{"is_live":true,"type":"regular","timestamp":1}}} |         clustering row | {"val":"1"}
  0 |       row-cache |                3 |                 |                                                                                         null |          partition end |        null

(3 rows)
```

- Both `n1` and `n2` still hold a reference to the tombstone in the row-cache. 
- `pk=0` is alive. 

Reading from cache and disk return inconsistent results. From `n1` (which still holds the data as alive):

```cql
cqlsh> CONSISTENCY LOCAL_ONE; SELECT * FROM k.t WHERE pk=0 BYPASS CACHE; SELECT * FROM k.t WHERE pk=0;
Consistency level set to LOCAL_ONE.

 pk | val
----+-----
  0 |   1

(1 rows)

 pk | val
----+-----

(0 rows)
cqlsh> 
```

The same happens to `n2` after read-repair.

In other words, the data can only be read until the row-cache evicts the tombstone. Considering the key is popular, the key will  forever be unreadable.

Even though this is not a "classic" data resurrection situation (we expected to be able to read the row in this example), this  could delay one to find out that their data was resurrected, which in turn would cause the "resurrected" row to remain alive on disk for an indefinite amount of time.

**Comment by mykaul** [comment]
@bhalevy - I assume this falls under your team.

**Comment by bhalevy** [comment]
@tgrabiec is there an open issue about that already?

**Comment by avikivity** [comment]
> Insert pk=0 again using an out of order write. In this example, the key timestamp is lower than on the delete:

The behavior after this is undefined. The system is allowed to keep expired tombstones indefinitely, and differently on different nodes. The rules are only about how long it is required to hole unexpired tombstones.

If you deliberately generate out of order writes, increase propagation_delay_in_seconds to match. The goal of the mutation timestamp is to resolve concurrent writes, not to allow the user to play clever tricks.

**Comment by bhalevy** [comment]
I find the inconsistency between the cache and disk the problem here, not the exposure of backdated data

**Comment by avikivity** [comment]
> I find the inconsistency between the cache and disk the problem here, not the exposure of backdated data

It's not a real inconsistency. An expired tombstone is expired. It can be discarded at will. So why is it important if it's kept in cache for longer or shorter periods than disk?

**Comment by tgrabiec** [comment]
The meaning of propagation_delay_in_seconds is that all writes with lower timestamps than at the start of the repair will propagate. gc_grace_seconds is similar - that all writes with lower timestamps propagate. The scenario violates that, so it's invalid. We cannot guarantee consistency in that case.

**Comment by raphaelsc** [comment]
I bumped into a scenario today where cache is completely empty, memtable and sstable are the only data sources.

Fact:
On read path, row cache applies compacting reader on sstable reader (only), so an expired tombstone read from a sstable can be purged (not from disk, but from the request) before it has chance to be merged with deleted data in memtable.

How it could be reproduced:
1) n1 has row @ t1 written to memtable of a table that has very low write activity
2) n1 misses the tombstone @ t2
3) repair runs and writes that tombstone @ t2 directly into a sstable in n1
4) with repair gc mode, that tombstone @ t2 gets expired after 1h (the default propagation delay)
5) since the table has low activity, row @ t1 still sits in memtable
6) on read path, compacting reader purges the tombstone @ t2 read from the sstable, before it's pushed to mutation merger, so the row @ t1 in memtable is returned to the client.

With bypass cache, there's no compacting reader so the expired tombstone is read from disk and merged with memtable. So no live data is returned.

I realized that compacting reader is ignoring the memtable content when calculating max purgeable timestamp (only commitlog content is taken into account on read path today). I am thinking it should look into memtable timestamp like compaction does, as compaction does it to avoid a unlikely but possible data resurrection (steps above show how it can happen). Alternatively the compacting reader wraps both the memtable and sstable data sources.

**Comment by raphaelsc** [comment]
> > I bumped into a scenario today where cache is completely empty, memtable and sstable are the only data sources.
> > Fact: On read path, table applies compacting reader on sstable reader (only), so an expired tombstone read from a sstable can be purged (not from disk, but from the request) before it has chance to be merged with deleted data in memtable.
> > How it could be reproduced:
> > 
> > 1. n1 has row @ t1 written to memtable of a table that has very low write activity
> > 2. n1 misses the tombstone @ t2
> > 3. repair runs and writes that tombstone @ t2 directly into a sstable in n1
> > 4. with repair gc mode, that tombstone @ t2 gets expired after 1h (the default propagation delay)
> > 5. since the table has low activity, row @ t1 still sits in memtable
> > 6. on read path, compacting reader purges the tombstone @ t2 read from the sstable, before it's pushed to mutation merger, so the row @ t1 in memtable is returned to the client.
> > 
> > With bypass cache, there's no compacting reader so the expired tombstone is read from disk and merged with memtable. So no live data is returned.
> > I realized that compacting reader is ignoring the memtable content when calculating max purgeable timestamp (only commitlog content is taken into account on read path today). I am thinking it should look into memtable timestamp like compaction does, as compaction does it to avoid a unlikely but possible data resurrection (steps above show how it can happen). Alternatively the compacting reader wraps both the memtable and sstable data sources.
> 
> That's [#1745](https://github.com/scylladb/scylladb/issues/1745), which was supposedly fixed.

So looks like we now have: "Data in memtable is not considered when deciding whether expired tombstone can be discarded during **read** <del>sstable compaction</del>". Will open an issue so we can keep track of it.

**Comment by raphaelsc** [comment]
> > > I bumped into a scenario today where cache is completely empty, memtable and sstable are the only data sources.
> > > Fact: On read path, table applies compacting reader on sstable reader (only), so an expired tombstone read from a sstable can be purged (not from disk, but from the request) before it has chance to be merged with deleted data in memtable.
> > > How it could be reproduced:
> > > 
> > > 1. n1 has row @ t1 written to memtable of a table that has very low write activity
> > > 2. n1 misses the tombstone @ t2
> > > 3. repair runs and writes that tombstone @ t2 directly into a sstable in n1
> > > 4. with repair gc mode, that tombstone @ t2 gets expired after 1h (the default propagation delay)
> > > 5. since the table has low activity, row @ t1 still sits in memtable
> > > 6. on read path, compacting reader purges the tombstone @ t2 read from the sstable, before it's pushed to mutation merger, so the row @ t1 in memtable is returned to the client.
> > > 
> > > With bypass cache, there's no compacting reader so the expired tombstone is read from disk and merged with memtable. So no live data is returned.
> > > I realized that compacting reader is ignoring the memtable content when calculating max purgeable timestamp (only commitlog content is taken into account on read path today). I am thinking it should look into memtable timestamp like compaction does, as compaction does it to avoid a unlikely but possible data resurrection (steps above show how it can happen). Alternatively the compacting reader wraps both the memtable and sstable data sources.
> > 
> > 
> > That's [#1745](https://github.com/scylladb/scylladb/issues/1745), which was supposedly fixed.
> 
> So looks like we now have: "Data in memtable is not considered when deciding whether expired tombstone can be discarded during **read** ~sstable compaction~". Will open an issue so we can keep track of it.

Opened https://github.com/scylladb/scylladb/issues/20916

**Comment by mykaul** [comment]
https://github.com/scylladb/scylladb/issues/20916 is fixed and backported to 6.0, 6.1, 6.2.
